File: 990/990.cs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/990.py

using System;
using System.Collections.Generic;

class Program
{
    const long MOD = 1_000_000_007L;
    const int MAX_N = 50;
    static readonly int MAX_TERMS = (MAX_N + 1) / 2;
    const int MAX_CARRY = 25;
    const int CARRY_MIN = -MAX_CARRY;
    const int CARRY_MAX = MAX_CARRY;

    static long[][] BINOM;
    static long[][][] SUM_TABLES;
    static Dictionary<(int, int, int), List<(int, int, int, long)>> transCache =
        new Dictionary<(int, int, int), List<(int, int, int, long)>>();

    static long[][] BuildBinom(int limit)
    {
        var comb = new long[limit + 1][];
        for (int n = 0; n <= limit; n++)
        {
            comb[n] = new long[limit + 1];
            comb[n][0] = 1;
            comb[n][n] = 1;
            for (int k = 1; k < n; k++)
                comb[n][k] = (comb[n - 1][k - 1] + comb[n - 1][k]) % MOD;
        }
        return comb;
    }

    static long[] ConvolveSmall(long[] poly, int width)
    {
        var outArr = new long[poly.Length + width - 1];
        for (int i = 0; i < poly.Length; i++)
        {
            if (poly[i] == 0) continue;
            for (int digit = 0; digit < width; digit++)
                outArr[i + digit] = (outArr[i + digit] + poly[i]) % MOD;
        }
        return outArr;
    }

    static long[][][] BuildSumTables(int limit)
    {
        var ways0to9 = new long[limit + 1][];
        ways0to9[0] = new long[] { 1L };
        for (int p = 1; p <= limit; p++)
            ways0to9[p] = ConvolveSmall(ways0to9[p - 1], 10);
        var tables = new long[limit + 1][][];
        for (int p = 0; p <= limit; p++)
        {
            tables[p] = new long[limit + 1][];
            tables[p][0] = ways0to9[p];
            for (int q = 1; q <= limit; q++)
                tables[p][q] = ConvolveSmall(tables[p][q - 1], 9);
        }
        return tables;
    }

    static List<(int, int, int, long)> GetTransitions(int activeLeft, int activeRight, int carry)
    {
        var key = (activeLeft, activeRight, carry);
        if (transCache.TryGetValue(key, out var cached)) return cached;
        if (activeLeft == 0 && activeRight == 0)
        {
            transCache[key] = new List<(int, int, int, long)>();
            return transCache[key];
        }
        var result = new List<(int, int, int, long)>();
        for (int nextLeft = 0; nextLeft <= activeLeft; nextLeft++)
        {
            long chooseLeft = BINOM[activeLeft][nextLeft];
            int endingLeft = activeLeft - nextLeft;
            for (int nextRight = 0; nextRight <= activeRight; nextRight++)
            {
                long chooseTerms = (chooseLeft * BINOM[activeRight][nextRight]) % MOD;
                int continuing = nextLeft + nextRight;
                int ending = (activeLeft - nextLeft) + (activeRight - nextRight);
                long[] counts = SUM_TABLES[continuing][ending];
                long baseVal = (long)(-carry - endingLeft + 9 * activeRight);
                for (int nextCarry = CARRY_MIN; nextCarry <= CARRY_MAX; nextCarry++)
                {
                    long idx = 10L * nextCarry + baseVal;
                    if (idx >= 0 && idx < counts.Length)
                    {
                        long ways = counts[(int)idx];
                        if (ways != 0)
                        {
                            long weight = (chooseTerms * ways) % MOD;
                            result.Add((nextLeft, nextRight, nextCarry, weight));
                        }
                    }
                }
            }
        }
        transCache[key] = result;
        return result;
    }

    static long Solve(int limit)
    {
        var dp = new Dictionary<(int, int, int), long>[limit + 1];
        for (int i = 0; i <= limit; i++)
            dp[i] = new Dictionary<(int, int, int), long>();
        for (int leftTerms = 1; leftTerms <= MAX_TERMS; leftTerms++)
        {
            for (int rightTerms = 1; rightTerms <= MAX_TERMS; rightTerms++)
            {
                int baseLength = leftTerms + rightTerms - 1;
                if (baseLength <= limit)
                {
                    var state = (leftTerms, rightTerms, 0);
                    dp[baseLength].TryGetValue(state, out long cur);
                    dp[baseLength][state] = (cur + 1) % MOD;
                }
            }
        }
        long answer = 0;
        for (int usedLength = 0; usedLength <= limit; usedLength++)
        {
            var current = dp[usedLength];
            if (current.Count == 0) continue;
            if (current.TryGetValue((0, 0, 0), out long zeroVal))
                answer = (answer + zeroVal) % MOD;
            var entries = new List<((int, int, int), long)>(current);
            foreach (var (state, waysSoFar) in entries)
            {
                if (waysSoFar == 0) continue;
                var (activeLeft, activeRight, carry) = state;
                if (activeLeft == 0 && activeRight == 0) continue;
                int nextLength = usedLength + activeLeft + activeRight;
                if (nextLength > limit) continue;
                var bucket = dp[nextLength];
                foreach (var (nextLeft, nextRight, nextCarry, weight) in GetTransitions(activeLeft, activeRight, carry))
                {
                    var newState = (nextLeft, nextRight, nextCarry);
                    bucket.TryGetValue(newState, out long cur2);
                    bucket[newState] = (cur2 + waysSoFar * weight) % MOD;
                }
            }
        }
        return answer;
    }

    static void RunSelfChecks()
    {
        System.Diagnostics.Debug.Assert(Solve(3) == 9);
        System.Diagnostics.Debug.Assert(Solve(5) == 171);
        System.Diagnostics.Debug.Assert(Solve(7) == 4878);
    }

    static void Main()
    {
        BINOM = BuildBinom(MAX_TERMS);
        SUM_TABLES = BuildSumTables(2 * MAX_TERMS);
        RunSelfChecks();
        Console.WriteLine(Solve(50));
    }
}

File: 990/990.rs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/990.py

use std::collections::HashMap;

const MOD: u64 = 1_000_000_007;
const MAX_N: usize = 50;
const MAX_TERMS: usize = (MAX_N + 1) / 2;
const MAX_CARRY: i32 = 25;
const CARRY_MIN: i32 = -MAX_CARRY;
const CARRY_MAX: i32 = MAX_CARRY;

fn build_binom(limit: usize) -> Vec<Vec<u64>> {
    let mut comb = vec![vec![0u64; limit + 1]; limit + 1];
    for n in 0..=limit {
        comb[n][0] = 1;
        comb[n][n] = 1;
        for k in 1..n {
            comb[n][k] = (comb[n - 1][k - 1] + comb[n - 1][k]) % MOD;
        }
    }
    comb
}

fn convolve_small(poly: &[u64], width: usize) -> Vec<u64> {
    let mut out = vec![0u64; poly.len() + width - 1];
    for (i, &value) in poly.iter().enumerate() {
        if value == 0 {
            continue;
        }
        for digit in 0..width {
            out[i + digit] = (out[i + digit] + value) % MOD;
        }
    }
    out
}

fn build_sum_tables(limit: usize) -> Vec<Vec<Vec<u64>>> {
    let mut ways_0_to_9: Vec<Vec<u64>> = vec![vec![]; limit + 1];
    ways_0_to_9[0] = vec![1];
    for p in 1..=limit {
        ways_0_to_9[p] = convolve_small(&ways_0_to_9[p - 1].clone(), 10);
    }
    let mut tables: Vec<Vec<Vec<u64>>> = vec![vec![vec![]; limit + 1]; limit + 1];
    for p in 0..=limit {
        tables[p][0] = ways_0_to_9[p].clone();
        for q in 1..=limit {
            let prev = tables[p][q - 1].clone();
            tables[p][q] = convolve_small(&prev, 9);
        }
    }
    tables
}

fn transitions(
    active_left: usize,
    active_right: usize,
    carry: i32,
    binom: &Vec<Vec<u64>>,
    sum_tables: &Vec<Vec<Vec<u64>>>,
    cache: &mut HashMap<(usize, usize, i32), Vec<(usize, usize, i32, u64)>>,
) -> Vec<(usize, usize, i32, u64)> {
    let key = (active_left, active_right, carry);
    if let Some(cached) = cache.get(&key) {
        return cached.clone();
    }
    if active_left == 0 && active_right == 0 {
        cache.insert(key, vec![]);
        return vec![];
    }
    let mut result: Vec<(usize, usize, i32, u64)> = Vec::new();
    for next_left in 0..=active_left {
        let choose_left = binom[active_left][next_left];
        let ending_left = active_left - next_left;
        for next_right in 0..=active_right {
            let choose_terms = (choose_left * binom[active_right][next_right]) % MOD;
            let continuing = next_left + next_right;
            let ending = (active_left - next_left) + (active_right - next_right);
            let counts = &sum_tables[continuing][ending];
            let base: i64 = -(carry as i64) - (ending_left as i64) + 9 * (active_right as i64);
            for next_carry in CARRY_MIN..=CARRY_MAX {
                let index = 10 * (next_carry as i64) + base;
                if index >= 0 && (index as usize) < counts.len() {
                    let ways = counts[index as usize];
                    if ways != 0 {
                        let weight = (choose_terms * ways) % MOD;
                        result.push((next_left, next_right, next_carry, weight));
                    }
                }
            }
        }
    }
    cache.insert(key, result.clone());
    result
}

fn solve(
    limit: usize,
    binom: &Vec<Vec<u64>>,
    sum_tables: &Vec<Vec<Vec<u64>>>,
    trans_cache: &mut HashMap<(usize, usize, i32), Vec<(usize, usize, i32, u64)>>,
) -> u64 {
    let mut dp: Vec<HashMap<(usize, usize, i32), u64>> = vec![HashMap::new(); limit + 1];

    for left_terms in 1..=MAX_TERMS {
        for right_terms in 1..=MAX_TERMS {
            let base_length = left_terms + right_terms - 1;
            if base_length <= limit {
                let state = (left_terms, right_terms, 0i32);
                let entry = dp[base_length].entry(state).or_insert(0);
                *entry = (*entry + 1) % MOD;
            }
        }
    }

    let mut answer = 0u64;
    for used_length in 0..=limit {
        let current: Vec<((usize, usize, i32), u64)> =
            dp[used_length].iter().map(|(&k, &v)| (k, v)).collect();
        if current.is_empty() {
            continue;
        }
        if let Some(&v) = dp[used_length].get(&(0, 0, 0)) {
            answer = (answer + v) % MOD;
        }
        for ((active_left, active_right, carry), ways_so_far) in current {
            if ways_so_far == 0 || (active_left == 0 && active_right == 0) {
                continue;
            }
            let next_length = used_length + active_left + active_right;
            if next_length > limit {
                continue;
            }
            let trans = transitions(active_left, active_right, carry, binom, sum_tables, trans_cache);
            for (next_left, next_right, next_carry, weight) in trans {
                let state = (next_left, next_right, next_carry);
                let entry = dp[next_length].entry(state).or_insert(0);
                *entry = (*entry + ways_so_far * weight) % MOD;
            }
        }
    }
    answer
}

fn run_self_checks(
    binom: &Vec<Vec<u64>>,
    sum_tables: &Vec<Vec<Vec<u64>>>,
    trans_cache: &mut HashMap<(usize, usize, i32), Vec<(usize, usize, i32, u64)>>,
) {
    assert_eq!(solve(3, binom, sum_tables, trans_cache), 9);
    assert_eq!(solve(5, binom, sum_tables, trans_cache), 171);
    assert_eq!(solve(7, binom, sum_tables, trans_cache), 4878);
}

fn main() {
    let binom = build_binom(MAX_TERMS);
    let sum_tables = build_sum_tables(2 * MAX_TERMS);
    let mut trans_cache: HashMap<(usize, usize, i32), Vec<(usize, usize, i32, u64)>> = HashMap::new();
    run_self_checks(&binom, &sum_tables, &mut trans_cache);
    println!("{}", solve(50, &binom, &sum_tables, &mut trans_cache));
}
